### PR TITLE
Fixes #22381 - Use in for parameters and arguments

### LINKED ISSUE #21981: Remove specialcasing for Span and ReadOnlySpan before shipping
We currently assume, regardless of attributes, that Span and ReadOnlySpan are
- ref-like structs
- readonly structs

This is done as a bootstrapping measure to mitigate the situation where some repos have old compilers and some implementations of span types do not have expected annotation.

We should remove the special-casing when before shipping to not set a compat precedent where annotations are unnecessary on span types.

### LINKED ISSUE #22381: Use `in` for parameters and arguments
From LDM discussion today:

1) **parameters should be declared with `in` instead of `ref readonly`**

2) plain method call with consider both byval and `in` candidates, and continue producing an error if ambiguity
3) `in` at call site guarantees the `in` candidate is picked and no copy is made
4) async spilling should never make a copy if `in` (an error should be produced in this case)

This bug is about part 1 

Other parts are tracked in https://github.com/dotnet/roslyn/issues/22387

### PR DESCRIPTION
**Pull request by OmarTawfik**
Fixes #22381 - Use `in` for parameters and arguments

### CODE REVIEW

File: src/Compilers/CSharp/Test/Emit/CodeGen/CodeGenRefReadOnlyParametersTests.cs [hunk 1]
```diff
@@ -13,15 +13,15 @@
 namespace Microsoft.CodeAnalysis.CSharp.UnitTests
 {
     [CompilerTrait(CompilerFeature.ReadOnlyReferences)]
-    public class CodeGenRefReadOnlyParametersTests : CompilingTestBase
```
Comment (OmarTawfik) [inline src/Compilers/CSharp/Test/Emit/CodeGen/CodeGenRefReadOnlyParametersTests.cs]: >CodeGenRefReadOnlyParametersTests [](start = 17, length = 33)

nit: rename the file before merging.

File: src/VisualStudio/Core/Def/Implementation/ChangeSignature/ChangeSignatureDialogViewModel.cs [hunk 2]
```diff
@@ -467,8 +467,8 @@ public string Modifier
                             return "out";
                         case RefKind.Ref:
                             return "ref";
-                        case RefKind.RefReadOnly:
-                            return "ref readonly";
+                        case RefKind.In:
+                            return "in";
```
Comment (OmarTawfik) [inline src/VisualStudio/Core/Def/Implementation/ChangeSignature/ChangeSignatureDialogViewModel.cs]: nit: break on `None`, and throw otherwise.

**Comment by OmarTawfik** [comment]
cc @VSadov @jcouv @dotnet/roslyn-compiler

File: src/Compilers/CSharp/Portable/Symbols/Source/SourceOrdinaryMethodSymbol.cs [hunk 3]
```diff
@@ -226,7 +226,7 @@ private void MethodChecks(MethodDeclarationSyntax syntax, Binder withTypeParamsB
                 {
                     diagnostics.Add(ErrorCode.ERR_RefExtensionMustBeValueTypeOrConstrainedToOne, location, Name);
                 }
-                else if (parameter0RefKind == RefKind.RefReadOnly && parameter0Type.TypeKind != TypeKind.Struct)
+                else if (parameter0RefKind == RefKind.In && parameter0Type.TypeKind != TypeKind.Struct)
                 {
                     diagnostics.Add(ErrorCode.ERR_RefReadOnlyExtensionMustBeValueType, location, Name);
```
Comment (VSadov) [inline src/Compilers/CSharp/Portable/Symbols/Source/SourceOrdinaryMethodSymbol.cs]: Should this be `ERR_InExtensionMethod. . . `?

File: src/Compilers/Core/Portable/PublicAPI.Unshipped.txt [hunk 4]
```diff
@@ -131,6 +131,7 @@ Microsoft.CodeAnalysis.OperationKind.VariableDeclaration = 1030 -> Microsoft.Cod
 Microsoft.CodeAnalysis.OperationKind.VariableDeclarationStatement = 3 -> Microsoft.CodeAnalysis.OperationKind
 Microsoft.CodeAnalysis.OperationKind.YieldBreakStatement = 12 -> Microsoft.CodeAnalysis.OperationKind
 Microsoft.CodeAnalysis.OperationKind.YieldReturnStatement = 16 -> Microsoft.CodeAnalysis.OperationKind
+Microsoft.CodeAnalysis.RefKind.In = 3 -> Microsoft.CodeAnalysis.RefKind
 Microsoft.CodeAnalysis.RefKind.RefReadOnly = 3 -> Microsoft.CodeAnalysis.RefKind
```
Comment (jcouv) [inline src/Compilers/Core/Portable/PublicAPI.Unshipped.txt]: Should RefKind.RefReadonly be removed? Or is that covered by Vlad's PR?

Comment (jcouv) [reply src/Compilers/Core/Portable/PublicAPI.Unshipped.txt]: Nevermind, we need it for returns...

---
In reply to: [141945392](https://github.com/dotnet/roslyn/pull/22424#discussion_r141945392) [](ancestors = 141945392)

**Comment by dpoeschl** [comment]
Tagging @dotnet/roslyn-ide as an FYI.

**Comment by jcouv** [comment]
ubuntu_14_debug_prtest  failed with error below. ([details](https://ci.dot.net/job/dotnet_roslyn/job/master/job/ubuntu_14_debug_prtest/6668/))

`error CS0234: The type or namespace name 'Loader' does not exist in the namespace 'System.Runtime' (are you missing an assembly reference?) [/mnt/j/workspace/dotnet_roslyn/master/ubuntu_14_debug_prtest/src/Scripting/Core/Scripting.csproj]`

FYI @dotnet/roslyn-infrastructure I'll re-run

**Comment by jcouv** [comment]
test ubuntu_14_debug_prtest please

**Comment by jasonmalinowski** [comment]
@jcouv @brettfo Do we have something tracking that as a flaky failure? "missing type" seems unlikely as a flaky anything.

**Comment by VSadov** [comment]
I will merge this since it is passing all tests and I need to rebase onto this change.

File: src/Compilers/CSharp/Portable/Symbols/Source/SourceOrdinaryMethodSymbol.cs (hunk 3 above)
Comment (OmarTawfik) [reply src/Compilers/CSharp/Portable/Symbols/Source/SourceOrdinaryMethodSymbol.cs]: Yes. I'll rename it in a later PR.

---
In reply to: [141921251](https://github.com/dotnet/roslyn/pull/22424#discussion_r141921251) [](ancestors = 141921251)

**Pull request merged**